### PR TITLE
PLAT-189154 [CMLE-watsonX] onboard CMLE using watsonX

### PR DESCRIPTION
**Pull request by enavitaniuc**
## Description

Introduce CMLE notebooks on top of watsonx

## Motivation and Context

As part of this PR we're introducing notebooks to interact with AEP from watsonX platform, this would enable another cloud ML platform to integrate and used with AEP

## How Has This Been Tested?

Run notebooks 3 thru 5 on WatsonX as per provided README.

## Screenshots (if appropriate):
Screenshots are included in PR.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: conf/config.ini [hunk 1]
```diff
@@ -5,7 +5,7 @@ dataset_id=
 featurized_dataset_id=
 dataflow_id=
 scoring_dataset_id=
-environment=prod
+environment=stage
```
Comment (vijeth8) [inline conf/config.ini:8]: why is this changed to stage

File: notebooks/watsonx/Week3Notebook_watsonx.ipynb [hunk 2]
```diff
@@ -0,0 +1,1572 @@
...
+   "metadata": {
+    "id": "2dddc758d151437c8930dae2c5fe1241"
+   },
+   "outputs": [],
+   "source": [
+    "wml_credentials = {\n",
+    "    \"instance_id\": \"openshift\",\n",
+    "    \"version\": \"4.8\",\n",
+    "    \"url\": \"https://cpd-cpd-instance.apps.p712zf6h.eastus2.aroapp.io\",\n",
+    "    \"username\": watson_username,\n",
```
Comment (vijeth8) [inline notebooks/watsonx/Week3Notebook_watsonx.ipynb:109]: Is it ok to hardcode url here

**Review by vijeth8 — APPROVED** [review]
Looks good, as long as notebooks are running e2e on watsonX.

File: conf/config.ini (hunk 1 above)
Comment (enavitaniuc) [reply conf/config.ini:8]: yeah can be reverted, normally there shouldn't be any value since customers are supposed to tweak these

File: notebooks/watsonx/Week3Notebook_watsonx.ipynb (hunk 2 above)
Comment (enavitaniuc) [reply notebooks/watsonx/Week3Notebook_watsonx.ipynb:109]: indeed, will have to add a note to ensure this argument here is tweaked according to customers provisioned endpoint here.

File: notebooks/watsonx/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+The steps below assume you have an account on watsonx, if not please ensure you have one. See [Watsonx](https://www.ibm.com/watsonx) for more information.
+
+### Setup
+
+* Navigate to watsonx `Project` tab and create new project ![Project](images/projects.png) (select `Create and empty Project`) Supply a name and description, save it.
+
+* Next step we need to pull `user_name` and `user_key` to be able to interact with watson programatically, navigate to `Profile and settings` top right corner, here you have access to user_name and key ![NameKey](images/user_name_key.png) 
+
+* Next store these values in config.ini (which you used through notebooks 1 and 2) under `[Watsonx]` section. Do not share them with anyone.
```
Comment (cmenguy) [inline notebooks/watsonx/README.md:17]: You can also add a link to the `config.ini` file.

File: notebooks/watsonx/Week3Notebook_watsonx.ipynb
Comment (cmenguy) [inline notebooks/watsonx/Week3Notebook_watsonx.ipynb:1]: Handling the DLZ backend would be better done in `aepp` but for now this is fine.

Comment (cmenguy) [inline notebooks/watsonx/Week3Notebook_watsonx.ipynb:1]: Typo on `trainig_data_connection`

File: conf/config.ini [hunk 4]
```diff
@@ -28,3 +28,7 @@ s3_bucket_name=
 s3_prefix=
 cfn_stack_id=
 
+[Watsonx]
+watson_username=
+watson_apikey=
+
```
Comment (cmenguy) [inline conf/config.ini:34]: The WatsonX URL and instance ID should also come from the config, since customers may not be on Openshift and will definitely not have the same endpoint URL as internally.

File: notebooks/watsonx/Week3Notebook_watsonx.ipynb
Comment (cmenguy) [inline notebooks/watsonx/Week3Notebook_watsonx.ipynb:1]: Make sure to remove the WatsonX URL from the notebook and instead put it in the config.

File: notebooks/watsonx/Week5Notebook_watsonx.ipynb
Comment (cmenguy) [inline notebooks/watsonx/Week5Notebook_watsonx.ipynb:1]: Do we need notebook 5 here? I don't think it does anything specific to WatsonX.

**Review by cmenguy — APPROVED** [review]
LGTM but make sure to change the WatsonX URL to come from the config. Other than that minor comments, thanks for doing that!

File: notebooks/watsonx/Week3Notebook_watsonx.ipynb
Comment (enavitaniuc) [reply notebooks/watsonx/Week3Notebook_watsonx.ipynb:1]: we would have to migrate quite some code to aepp, additionally all the spark configs would have to be moved to aepp.

File: conf/config.ini (hunk 4 above)
Comment (enavitaniuc) [reply conf/config.ini:34]: I wonder if version should be moved too, where can I get an understanding what are the expected values for other solutions?

File: notebooks/watsonx/Week5Notebook_watsonx.ipynb
Comment (enavitaniuc) [reply notebooks/watsonx/Week5Notebook_watsonx.ipynb:1]: true, however this way consumers of watsonx would have to flip between local to cloud to back to local for last step (including downloading config.ini from the cloud platform) while I agree with you there is nothing specific to watsonx in this notebook it feels more straightforward to finish the week5 on the watsonx, WDYT?

File: conf/config.ini (hunk 4 above)
Comment (enavitaniuc) [reply conf/config.ini:34]: based on: https://www.ibm.com/docs/en/cloud-paks/cp-data/4.8.x?topic=assets-authenticating-programmatic-access

it appears there is no easy way to figure on the fly exact instanceId and version, one should check the exact cloud park version the is provisioned on. will add this info to the readme.
